geo/src/algorithm/haversine_closest_point.rs

```diff
@@ -0,0 +1,579 @@
+use crate::{haversine_distance::HaversineDistance, Bearing};
+use crate::{Closest, Contains};
+use crate::{CoordsIter, GeoFloat, HaversineDestination, Point, MEAN_EARTH_RADIUS};
+use geo_types::{
+    Coord, Geometry, GeometryCollection, Line, LineString, MultiLineString, MultiPoint,
+    MultiPolygon, Polygon, Rect, Triangle,
+};
+
+use num_traits::FromPrimitive;
+
+/// Closest point on a geometry from a point using great circles.
+///
+/// https://edwilliams.org/avform147.htm#XTE
+///
+pub trait HaversineClosestPoint<T>
+where
+    T: GeoFloat + FromPrimitive,
+{
+    fn haversine_closest_point(&self, from: &Point<T>) -> Closest<T>;
+}
+
+// Implement for references as well as types
+impl<'a, T, G> HaversineClosestPoint<T> for &'a G
+where
+    G: HaversineClosestPoint<T>,
+    T: GeoFloat + FromPrimitive,
+{
+    fn haversine_closest_point(&self, from: &Point<T>) -> Closest<T> {
+        (*self).haversine_closest_point(from)
+    }
+}
+
+impl<T> HaversineClosestPoint<T> for Point<T>
+where
+    T: GeoFloat + FromPrimitive,
+{
+    fn haversine_closest_point(&self, pt: &Point<T>) -> Closest<T> {
+        if self == pt {
+            Closest::Intersection(*self)
+        } else {
+            Closest::SinglePoint(*pt)
+        }
+    }
+}
+
+impl<T> HaversineClosestPoint<T> for Coord<T>
+where
+    T: GeoFloat + FromPrimitive,
+{
+    fn haversine_closest_point(&self, pt: &Point<T>) -> Closest<T> {
+        Point::from(*self).haversine_closest_point(pt)
+    }
+}
+
+impl<T> HaversineClosestPoint<T> for Line<T>
+where
+    T: GeoFloat + FromPrimitive,
+{
+    fn haversine_closest_point(&self, from: &Point<T>) -> Closest<T> {
+        // There is a question how small this distance should be. This is really tiny (~10e-7 meters or smaller),
+        // when working on earth, but it is arbitrary.
+        const DISTANCE_EPSILON: f64 = 1e-12;
```

Comment:
Can add to the documentation that non-intersecting geometries may be considered to intersect if they are very close within `T::EPSILON`.

I am paranoid that this might cause someone some agony, but I don't have a better solution and would love to see this get merged!